File: leetcode2/DungeonGame/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 112ms 33.05% 14.8MB 65.25%
# dynamic programming
# O(HW) O(HW)
class Solution:
	def calculateMinimumHP(self, G: List[List[int]]) -> int:
		h, w = len(G), len(G[0])
		C = [[float("inf")] * (w + 1) for _ in range(h + 1)]
		C[h][w-1], C[h-1][w] = 1, 1
		for y in range(h-1, -1, -1):
			for x in range(w-1, -1, -1):
				hp = min(C[y+1][x], C[y][x+1]) - G[y][x]
				if hp <= 0:
					hp = 1
				C[y][x] = hp
		return C[0][0]

File: leetcode2/DungeonGame/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <vector>
#include <cstdlib>

using namespace std;

// -2  -3  3 
// -5 -10  1 
// 10  30 -5 

//  7   5  2 M
//  6  11  5 M
//  1   1  6 1
//  M   M  1 M

// 20ms 12.89% 9MB 55.03%
// dynamic programming
// O(HW) O(HW)
class Solution {
 public:
  int calculateMinimumHP(vector<vector<int>>& G) {
    int h = G.size(), w = G[0].size();
    vector<vector<int>> C(h+1, vector<int>(w+1, INT_MAX));
    C[h][w-1] = 1;
    C[h-1][w] = 1;
    for (int y = h-1; y >= 0; --y) {
      for (int x = w-1; x >= 0; --x) {
        int need = min(C[y+1][x], C[y][x+1]) - G[y][x];
        if (need <= 0)
          need = 1;
        C[y][x] = need;
      }
    }
    return C[0][0];
  }
};
